compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -564,86 +565,127 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
             }
         }
     }
+}
 
-    /// Creates an incomplete operand containing the [`abi::Scalar`]s expected based
-    /// on the `layout` passed. This is for use with [`OperandRef::insert_field`]
-    /// later to set the necessary immediate(s).
-    ///
-    /// Returns `None` for `layout`s which cannot be built this way.
-    pub(crate) fn builder(
-        layout: TyAndLayout<'tcx>,
-    ) -> Option<OperandRef<'tcx, Result<V, abi::Scalar>>> {
-        // Uninhabited types are weird, because for example `Result<!, !>`
-        // shows up as `FieldsShape::Primitive` and we need to be able to write
-        // a field into `(u32, !)`. We'll do that in an `alloca` instead.
-        if layout.uninhabited {
-            return None;
-        }
+/// Each of these variants starts out as `Either::Right` when it's uninitialized,
+/// then setting the field changes that to `Either::Left` with the backend value.
+#[derive(Debug, Copy, Clone)]
+enum OperandValueBuilder<V> {
+    ZeroSized,
+    Immediate(Either<V, abi::Scalar>),
+    Pair(Either<V, abi::Scalar>, Either<V, abi::Scalar>),
+    /// SIMD vectors need special handling because they're the only case where
+    /// a type with a (non-ZST) `Memory`-ABI field can be `Scalar`-ABI.
```

Comment:
I don't understand this comment... `BackendRepr::Memory` and` BackendRepr::SimdVector` are mutually exclusive, how can a type be both?

"ABI" is such an overused term, it may be good to use some more precise term here if there is one.

Reply:
Ah, I missed the fact that "field" is the key term here. :)

Yeah that is kind of odd... might be worth changing? Haven't you been working on entirely forbidding field projections for SIMD vectors?